db/batch/kv_cache.go

```diff
@@ -6,96 +6,170 @@
-)
-
 type (
 	// KVStoreCache is a local cache of batched <k, v> for fast query
 	KVStoreCache interface {
 		// Read retrieves a record
-		Read(hash160 hash.Hash160) ([]byte, error)
+		Read(*KvCacheKey) ([]byte, error)
 		// Write puts a record into cache
-		Write(hash.Hash160, []byte)
+		Write(*KvCacheKey, []byte)
 		// WriteIfNotExist puts a record into cache only if it doesn't exist, otherwise return ErrAlreadyExist
-		WriteIfNotExist(hash.Hash160, []byte) error
+		WriteIfNotExist(*KvCacheKey, []byte) error
 		// Evict deletes a record from cache
-		Evict(hash.Hash160)
+		Evict(*KvCacheKey)
 		// Clear clear the cache
 		Clear()
 		// Clone clones the cache
 		Clone() KVStoreCache
 	}
 
+	// KvCacheKey is the key for 2D Map cache
+	KvCacheKey struct {
+		key1 string
+		key2 string
+	}
+
 	// kvCache implements KVStoreCache interface
 	kvCache struct {
-		cache   map[hash.Hash160][]byte // local cache of batched <k, v> for fast query
-		deleted map[hash.Hash160]struct{}
+		cache   map[string]map[string][]byte // local cache of batched <k, v> for fast query
+		deleted map[string]map[string]struct{}
 	}
 )
 
 // NewKVCache returns a KVCache
 func NewKVCache() KVStoreCache {
-	c := kvCache{
-		cache:   make(map[hash.Hash160][]byte),
-		deleted: make(map[hash.Hash160]struct{}),
+	return &kvCache{
+		cache:   make(map[string]map[string][]byte),
+		deleted: make(map[string]map[string]struct{}),
 	}
-	return &c
 }
 
 // Read retrieves a record
-func (c *kvCache) Read(k hash.Hash160) ([]byte, error) {
-	if v, ok := c.cache[k]; ok {
+func (c *kvCache) Read(key *KvCacheKey) ([]byte, error) {
+	if v := c.cacheGet(key); v != nil {
 		return v, nil
 	}
-	if _, ok := c.deleted[k]; ok {
+	if c.deletedGet(key) {
 		return nil, ErrAlreadyDeleted
 	}
 	return nil, ErrNotExist
 }
 
 // Write puts a record into cache
-func (c *kvCache) Write(k hash.Hash160, v []byte) {
-	c.cache[k] = v
-	delete(c.deleted, k)
+func (c *kvCache) Write(key *KvCacheKey, v []byte) {
+	c.cachePut(key, v)
+	c.deletedDelete(key)
 }
 
 // WriteIfNotExist puts a record into cache only if it doesn't exist, otherwise return ErrAlreadyExist
-func (c *kvCache) WriteIfNotExist(k hash.Hash160, v []byte) error {
-	if _, ok := c.cache[k]; ok {
+func (c *kvCache) WriteIfNotExist(key *KvCacheKey, v []byte) error {
+	if v := c.cacheGet(key); v != nil {
 		return ErrAlreadyExist
 	}
-	c.cache[k] = v
-	delete(c.deleted, k)
+	c.Write(key, v)
 	return nil
 }
 
 // Evict deletes a record from cache
-func (c *kvCache) Evict(k hash.Hash160) {
-	delete(c.cache, k)
-	c.deleted[k] = struct{}{}
+func (c *kvCache) Evict(key *KvCacheKey) {
+	c.cacheDelete(key)
+	c.deletedPut(key)
 }
 
 // Clear clear the cache
 func (c *kvCache) Clear() {
-	c.cache = nil
-	c.deleted = nil
-	c.cache = make(map[hash.Hash160][]byte)
-	c.deleted = make(map[hash.Hash160]struct{})
+	c.cacheClear()
+	c.deletedClear()
+	c.cache = make(map[string]map[string][]byte)
+	c.deleted = make(map[string]map[string]struct{})
```

Comment:
also the way you implement `clear()`, is to iterate and delete all keys, instead of simply set `c.cache = nil`
so these 2 lines are not needed, the 2 maps goes back to empty map after `clear()`